contracts/VestingEscrow.vy

```
@@ -0,0 +1,314 @@
+disabled_at: public(uint256)
+initialized: public(bool)
+
+owner: public(address)
+manager: public(address)
+
+
+@external
+def __init__(voting_addr: address, snapshot_addr: address):
+    """
+    @notice Initialize source contract implementation.
+    @param voting_addr Address of the Lido Voting contract
+    @param snapshot_addr Address of the Shapshot Delegate contract
+    """
+    # ensure that the original contract cannot be initialized
+    self.initialized = True
+    LIDO_VOTING_CONTRACT_ADDR = voting_addr
+    SNAPSHOT_DELEGATE_CONTRACT_ADDR = snapshot_addr
+
+
+@external
+@nonreentrant("lock")
+def initialize(
+    token: address,
+    amount: uint256,
+    recipient: address,
+    owner: address,
+    manager: address,
+    start_time: uint256,
+    end_time: uint256,
+    cliff_length: uint256,
+    voting_adapter_addr: address,
+) -> bool:
+    """
+    @notice Initialize the contract.
+    @dev This function is separate from `__init__` because of the factory pattern
+         used in `VestingEscrowFactory.deploy_vesting_contract`. It may be called
+         once per deployment.
+    @param token Address of the ERC20 token being distributed
+    @param amount Amount of the ERC20 token to be controleed by escrow
+    @param recipient Address to vest tokens for
+    @param owner Address of the vesting owner
+    @param manager Address of the vesting manager
+    @param start_time Epoch time at which token distribution starts
+    @param end_time Time until everything should be vested
+    @param cliff_length Duration after which the first portion vests
+    @param voting_adapter_addr VotingAdapter address
+    """
+    assert not self.initialized, "can only initialize once"
+    self.initialized = True
+
+    self.owner = owner
+    self.manager = manager
+    self.token = ERC20(token)
+    self.start_time = start_time
+    self.end_time = end_time
+    self.cliff_length = cliff_length
+
+    assert self.token.transferFrom(
```

in general case, `transfer From` implementation can silently fail inside, and assert would not fire.
See the issue https://github.com/vyperlang/vyper/issues/2202

As a workaround, use our _safeTransfer implementation from [Curve reward manager](https://github.com/lidofinance/curve-rewards-manager/blob/master/contracts/RewardsManager.vy#L181).
(also see the comment for `recover_erc20` method)